e2e/nomostest/portforwarder/port_forwarder.go

```diff
@@ -190,67 +196,155 @@ func (pf *PortForwarder) portForwardToPod(pod string, async bool) error {
-			return errors.Wrap(err, "failed to kill port forward process")
-		}
-	}
 	pf.logger.Infof("starting port-forward process for %s/%s", pf.ns, pf.deployment)
 	pod, err := pf.kubeClient.GetDeploymentPod(pf.deployment, pf.ns, pf.retryTimeout)
 	if err != nil {
 		return errors.Wrap(err, fmt.Sprintf("failed to get deployment pod for %s/%s", pf.ns, pf.deployment))
 	}
-	err = pf.portForwardToPod(pod.Name, async)
+	err = pf.portForwardToPod(pod.Name)
 	if err != nil {
 		return errors.Wrap(err, fmt.Sprintf("failed to start port forward for %s/%s", pf.ns, pf.deployment))
 	}
 	return nil
 }
 
-// start creates a port forward to a deployment's pod and keeps it open,
+// Start creates a port forward to a deployment's pod and keeps it open,
 // even if the pod is recreated.
-// Once ctx is cancelled, the port forward will be closed and will no longer be recreated
-func (pf *PortForwarder) start() error {
-	if err := pf.portForwardToDeployment(false); err != nil {
-		return err
+// Once ctx is cancelled, the port forward will be closed and will no longer be recreated.
+// Returns a channel. An error will be sent to the channel, and the channel will
+// be closed, when the PortForwarder stops.
+func (pf *PortForwarder) Start() chan error {
+	pf.mux.Lock()
+	defer pf.mux.Unlock()
+	// defend against the PortForwarder being started more than once
+	if pf.isStarted {
+		panic("PortForwarder already started")
 	}
+	pf.isStarted = true
 	// Restart the port forward if the process exits prematurely (e.g. due to pod eviction)
-	go func() {
-		for {
-			if err := pf.watcher.WatchForNotFound(kinds.Pod(), pf.pod, pf.ns, testwatcher.WatchTimeout(0), testwatcher.WatchContext(pf.ctx)); err != nil {
-				if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
-					// stop retrying, the subprocess will be killed with the context.
-					return
-				}
-				// Watch automatically retries, so this shouldn't happen unless
-				// there's a terminal error.
-				pf.logger.Infof("Watch exited, restarting port-forward for %s/%s: %v", pf.ns, pf.deployment, err)
-			}
-			if err := pf.portForwardToDeployment(true); err != nil {
+	errorCh := wrapWithChannel(func() error {
+		for { // loop forever until context is done or a terminal error
+			err := pf.startAndWait()
+			if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+				pf.logger.Infof("PortForwarder for pod %s/%s received done signal", pf.ns, pf.pod)
+				return nil
+			} else if err != nil && isTerminalError(err) {
+				err = errors.Wrapf(err, "PortForwarder for Pod %s/%s encountered a fatal error", pf.ns, pf.pod)
 				pf.logger.Info(err)
-				// this is a potentially fatal error since we failed to restart the port
-				// forward. we log the error and return in case the test is already done
-				// with the port forward. the test will otherwise error the next time it
-				// tries to use the local port.
-				return
+				return err
+			} else if err != nil {
```

Comment:
Can you pull this `if err != nil` out and make 2 layers of ifs? I think it would be more readable than this switch-like flow.

Reply:
sure

Reply:
done